amoro-ams/src/main/java/org/apache/amoro/server/catalog/DefaultCatalogManager.java

```diff
@@ -0,0 +1,178 @@
+import org.apache.amoro.shade.guava32.com.google.common.collect.Maps;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class DefaultCatalogManager extends PersistentBase implements CatalogManager {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultCatalogManager.class);
+  protected final Configurations serverConfiguration;
+
+  private final Map<String, ServerCatalog> serverCatalogMap = Maps.newConcurrentMap();
+
+  public DefaultCatalogManager(Configurations serverConfiguration) {
+    this.serverConfiguration = serverConfiguration;
+    listCatalogMetas()
+        .forEach(
+            c -> {
+              ServerCatalog serverCatalog =
+                  CatalogBuilder.buildServerCatalog(c, serverConfiguration);
+              serverCatalogMap.put(c.getCatalogName(), serverCatalog);
+            });
+    LOG.info("DefaultCatalogManager initialized, total catalogs: {}", serverCatalogMap.size());
+  }
+
+  @Override
+  public List<CatalogMeta> listCatalogMetas() {
+    return getAs(CatalogMetaMapper.class, CatalogMetaMapper::getCatalogs);
+  }
+
+  @Override
+  public CatalogMeta getCatalogMeta(String catalogName) {
+    return getCatalogMetaOptional(catalogName)
+        .orElseThrow(() -> new ObjectNotExistsException("Catalog " + catalogName));
+  }
+
+  private Optional<CatalogMeta> getCatalogMetaOptional(String catalogName) {
+    return Optional.ofNullable(
+        getAs(
+            CatalogMetaMapper.class,
+            catalogMetaMapper -> catalogMetaMapper.getCatalog(catalogName)));
+  }
+
+  @Override
+  public boolean catalogExist(String catalogName) {
+    return getCatalogMetaOptional(catalogName).isPresent();
+  }
+
+  @Override
+  public ServerCatalog getServerCatalog(String catalogName) {
+    Optional<CatalogMeta> catalogMeta = getCatalogMetaOptional(catalogName);
+    if (!catalogMeta.isPresent()) {
+      // remove if catalog is deleted
+      disposeCatalog(catalogName);
+      throw new ObjectNotExistsException("Catalog " + catalogName);
+    }
+    ServerCatalog serverCatalog =
+        serverCatalogMap.computeIfAbsent(
+            catalogName,
+            n -> CatalogBuilder.buildServerCatalog(catalogMeta.get(), serverConfiguration));
+    serverCatalog.reload();
+    return serverCatalog;
+  }
+
+  @Override
+  public InternalCatalog getInternalCatalog(String catalogName) {
+    ServerCatalog serverCatalog = getServerCatalog(catalogName);
+    if (serverCatalog == null) {
+      throw new ObjectNotExistsException("Catalog " + catalogName);
+    }
+    if (serverCatalog.isInternal()) {
+      return (InternalCatalog) serverCatalog;
+    }
+    throw new ObjectNotExistsException("Catalog " + catalogName + " is not internal catalog");
+  }
+
+  @Override
+  public List<ExternalCatalog> getExternalCatalogs() {
+    return listCatalogMetas().stream()
```

Comment:
We can improve the performance of this method by reducing some database calls.
Now it will select all the catalog metadata and then select each one of them individually, which should be improved.

Reply:
fixed